Update module.py as per #142306

grussdorian opened the pull request:
release notes: fx Issue  #142306
Minor work in Improve typing of args and kwargs with ParamSpec

**register_forward_hook**
    
      Key changes:
    
      1. Replace Any with Input/Output type vars for inputs/outputs
      2. Ensure the output type of the hook matches its input type
      3. Keep the Dict[str, Any] for kwargs as those are arbitrary
    
**register_forward_pre_hook**
    
      Key changes:
    
      1. Replace Any with Input type var for inputs
      2. Ensure the return type matches the input type structure
      3. Keep Dict[str, Any] for kwargs

grussdorian committed:
Improve typing of args and kwargs with ParamSpec #142306

1. register_forward_hook
    
      Key changes:
    
      1. Replace Any with Input/Output type vars for inputs/outputs
      2. Ensure the output type of the hook matches its input type
      3. Keep the Dict[str, Any] for kwargs as those are arbitrary
    
2. register_forward_pre_hook
    
      Key changes:
    
      1. Replace Any with Input type var for inputs
      2. Ensure the return type matches the input type structure
      3. Keep Dict[str, Any] for kwargs

grussdorian committed:
Merge pull request #1 from grussdorian/grussdorian-patch-1

Improve typing of args and kwargs with ParamSpec #142306

grussdorian commented:
release notes: fx

grussdorian commented:
@pytorchbot label "release notes: fx"

Skylion007 commented on torch/nn/modules/module.py:
```diff
@@ -47,7 +51,11 @@
 # the type of the subclass, not the looser type of `Module`.
 T = TypeVar("T", bound="Module")
-
+P = ParamSpec('P')
+R = TypeVar('R')
+Module_T = TypeVar('Module_T', bound='Module')
+Input = TypeVar('Input')
+Output = TypeVar('Output')
```
Nit: Want to Prefix this these TypeVar and Specs so they don't accidentally get exported:
```suggestion
_P = ParamSpec('_P')
_R = TypeVar('_R')
_Module_T = TypeVar('_Module_T', bound='Module')
_Input = TypeVar('_Input')
_Output = TypeVar('_Output')
```

Skylion007 commented on torch/nn/modules/module.py:
```diff
@@ -20,7 +20,11 @@
+from typing_extensions import (
+  Self,
+  ParamSpec,
+  TypeVar,
```
```suggestion
```

Skylion007 reviewed (changes requested):
Any reason to reimport from typing_extensions?

grussdorian commented:
Sorry reimported TypeVar  again, no need (overlooked that my bad)

grussdorian committed:
Update torch/nn/modules/module.py

removed double import

Co-authored-by: Aaron Gokaslan <aaronGokaslan@gmail.com>

grussdorian committed:
Update torch/nn/modules/module.py

measures against accidental export

Co-authored-by: Aaron Gokaslan <aaronGokaslan@gmail.com>

Skylion007 commented on torch/nn/modules/module.py:
```diff
@@ -1580,10 +1587,10 @@ def _maybe_warn_non_full_backward_hook(self, inputs, result, grad_fn):
         self,
         hook: Union[
-            Callable[[T, Tuple[Any, ...]], Optional[Any]],
+            Callable[[T, Tuple[Input, ...]], Optional[Tuple[Input, ...]]],
```
You also need to rename the TypeVars since you applied that change above

grussdorian replied:
wouldn't that be _Input instead of Input ? (Sorry for the stupid question)

Skylion007 replied:
Yes, also _Module_T is unused? How is it different from T?

grussdorian replied:
It's the same thing, however it's better to keep `T` as it's used by many other parts of the code. Even though we are following the _ naming convention

Skylion007 replied:
But it's not used anywhere else in the code, it's only defined and never referenced.

grussdorian replied:
Sorry for the confusion. Removed unnecessary `_Module_T`

grussdorian committed:
Update module.py

Resolved undefined names, Input  -> _Input and Output -> _Output and removed redundant _Module_T Type variable

Skylion007 commented on torch/nn/modules/module.py:
```diff
@@ -47,7 +50,10 @@
 # the type of the subclass, not the looser type of `Module`.
 T = TypeVar("T", bound="Module")
-
+_P = ParamSpec('_P')
```
_R isn't used at all either?

grussdorian replied:
Forgot to remove it, was changing return types before that and forgot to remove it
